// nimblegecko/views/oauth.js

var URL = require('url')
  , qs = require('querystring')
  , ng = require('nimblegecko')
  , _ = require('underscore')
  , OAUTH_CREDENTIALS = 'oAuthCredentials'


exports.callback = function(req, res, next) {

    var reqUrl = URL.parse(req.url)
      , oauth_verifier
      , oauth_credentials = ng.session.getObject(req, OAUTH_CREDENTIALS)

    // TODO: refactor this to use utils.checkRequiredOptions ?
    if(oauth_credentials === undefined ||
       oauth_credentials.oauth_token === undefined ||
       oauth_credentials.oauth_token_secret === undefined) {

        console.log('Creds: ')
        console.dir(oauth_credentials)
        ng.http.error(req, res, 'Error, no oauth_token / oauth_token_secret in session')
        return
    }

    oauth_verifier = qs.parse(reqUrl.query).oauth_verifier

    ng.oauth_tools.getOAuthAccessToken({

        oauth_token: oauth_credentials.oauth_token,
        oauth_token_secret: oauth_credentials.oauth_token_secret,
        oauth_verifier: oauth_verifier,

        next: function(error, oauth_token, oauth_token_secret, oauth_data) {

            if(error) {
                ng.http.error(req, res, error)
                return
            }

            ng.api.handleLogin({
                oauth_token: oauth_token,
                oauth_token_secret: oauth_token_secret,
                oauth_data: oauth_data,
                next: function(err, user) {

                    if (err) {
                        ng.http.error(req, res, err)
                        return
                    }

                    ng.session.setLoggedInUser(req, user)
                    ng.http.redirect(res, '/home')
                }
            })

        }

    })
}

exports.login = function(req, res, next) {

    ng.oauth_tools.getOAuthRequestToken(

        function(err, oauth_token, oauth_token_secret, results) {

            if(err) {
                ng.http.error(req, res, err)
            } else {
                ng.session.storeObject(req, OAUTH_CREDENTIALS,
                    {
                        oauth_token: oauth_token,
                        oauth_token_secret: oauth_token_secret
                    }
                )

                ng.oauth_tools.redirectToTwitterAuth(res, oauth_token)
            }
        }

    )
}
