// nimblegecko/oauth-tools.js
var oAuth = require('oauth').OAuth,
    URL = require('url'),
    util = require('util'),
    ng = require('nimblegecko')


function getAuth() {

    return new oAuth(
        'https://api.twitter.com/oauth/request_token',
        'https://api.twitter.com/oauth/access_token',
        ng.conf.oauth_api_key,
        ng.conf.oauth_secret,
        '1.0',
        null,
        'HMAC-SHA1'
    )

}


// TODO: Change that to use params (hostname, port) from request?
exports.getOAuthRequestToken = function(callback) {

    var callback_url = {
        protocol: 'http',
        hostname: ng.conf.callback_address,
        port: ng.conf.callback_port,
        pathname: ng.conf.oauth_callback_url
    }

    getAuth().getOAuthRequestToken({oauth_callback: URL.format(callback_url)}, callback)
}


exports.getOAuthAccessToken = function(params) {

    ng.utils.checkRequiredOptions(params, [
        'oauth_token',
        'oauth_token_secret',
        'oauth_verifier',
        'next'
    ])

    getAuth().getOAuthAccessToken(params.oauth_token, params.oauth_token_secret, params.oauth_verifier, params.next)
}


exports.redirectToTwitterAuth = function(res, oauth_token) {
    var url = URL.parse('http://api.twitter.com/oauth/authenticate')
    url.query = { oauth_token: oauth_token }
    ng.http.redirect(res, URL.format(url))
}
